bricklayers.py

```diff
@@ -1,143 +1,282 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see .
+#
+# Copyright (c) [2025] [Roman Tenger]
+
+import re
+import sys
+import logging
+import os
+import argparse
+from tempfile import NamedTemporaryFile
+import io
+from pathlib import Path
+
+class GCodeProcessor:
+    def __init__(self, layer_height, extrusion_multiplier):
+        self.layer_height = layer_height
+        self.extrusion_multiplier = extrusion_multiplier
+        self.current_layer = 0
+        self.current_z = 0.0
+        self.perimeter_type = None
+        self.perimeter_block_count = 0
+        self.inside_perimeter_block = False
+        self.z_shift = layer_height * 0.5
+        self.total_layers = 0
+        self.buffer_size = 1024 * 1024  # 1MB buffer size
+        self.printer_type = None
+        self.in_object = False
+        self.shifted_blocks = 0
+        self.perimeter_found = False
+
+    def detect_printer_type(self, lines):
+        """Detect printer type based on G-code features"""
+        logging.info("Starting printer type detection")
+        
+        for i, line in enumerate(lines):
+            if "; FEATURE:" in line:
+                logging.info(f"Detected Bambu/Orca printer from feature marker in line {i}: {line.strip()}")
+                return "bambu"
+            elif ";TYPE:" in line:
+                logging.info(f"Detected Prusa printer from type marker in line {i}: {line.strip()}")
+                return "prusa"
+        
+        logging.warning("No printer type markers found - defaulting to Prusa")
+        return "prusa"
+
+    def get_z_height_from_comment(self, line):
+        """Extract Z height from comment if present"""
+        if "; Z_HEIGHT:" in line:
+            match = re.search(r'; Z_HEIGHT: ([\d.]+)', line)
+            if match:
+                return float(match.group(1))
+        return None
+
+    def count_layers(self, input_file):
+        """Count layers using buffered reading"""
+        count = 0
+        with open(input_file, 'r') as f:
+            buffer = io.StringIO()
+            while True:
+                chunk = f.read(self.buffer_size)
+                if not chunk:
+                    break
+                buffer.write(chunk)
+                buffer.seek(0)
+                
+                for line in buffer:
+                    if line.startswith("G1 Z"):
+                        count += 1
+                        
+                buffer.seek(0)
+                buffer.truncate(0)
+        return count
+
+    def process_line(self, line, line_num, lines, is_last_layer):
```

Comment:
lots of branching logic in here -- often a sign a function is trying to do too many things

could think about splitting out helper functions for things like doing bambu vs prusa perimeters, applying the z shift, etc

that'll make it easier to read, and to test too